### PR TITLE
fix(iceberg): fix jni context class loader

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Try to resolve https://github.com/risingwavelabs/risingwave/issues/17296

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by wenym1** [comment]
I am thinking that shall we instead always set the context loader after we do attach_current_thread of jni?

**Comment by chenzl25** [comment]
> I am thinking that shall we instead always set the context loader after we do attach_current_thread of jni?

It depends on whether the method needs to use `ContextLoader`. Currently, I only find debezium and iceberg catalog will use it.

**Comment by wenym1** [comment]
> > I am thinking that shall we instead always set the context loader after we do attach_current_thread of jni?
> 
> It depends on whether the method needs to use `ContextClassLoader`. Currently, I only find debezium and iceberg catalog will use it.

Java libraries usually assume that the class loader is correctly set. So I think we'd better just always set the class loader when we do `attach_current_thread` to create a new `JniEnv`, so that we don't have to manually set the class loader one by one for these methods and can avoid similar exception thrown again when we use other java libraries in the future.

**Comment by chenzl25** [comment]
> > > I am thinking that shall we instead always set the context loader after we do attach_current_thread of jni?
> > 
> > 
> > It depends on whether the method needs to use `ContextClassLoader`. Currently, I only find debezium and iceberg catalog will use it.
> 
> Java libraries usually assume that the class loader is correctly set. So I think we'd better just always set the class loader when we do `attach_current_thread` to create a new `JniEnv`, so that we don't have to manually set the class loader one by one for these methods and can avoid similar exception thrown again when we use other java libraries in the future.

Sounds good. Do you have any idea how to implement it?

**Comment by wenym1** [comment]
> > > > I am thinking that shall we instead always set the context loader after we do attach_current_thread of jni?
> > > 
> > > 
> > > It depends on whether the method needs to use `ContextClassLoader`. Currently, I only find debezium and iceberg catalog will use it.
> > 
> > 
> > Java libraries usually assume that the class loader is correctly set. So I think we'd better just always set the class loader when we do `attach_current_thread` to create a new `JniEnv`, so that we don't have to manually set the class loader one by one for these methods and can avoid similar exception thrown again when we use other java libraries in the future.
> 
> Sounds good. Do you have any idea how to implement it?

We can unify all our call to `jvm.attach_current_thread()` to a method like
```
fn new_jvm_env(jvm: &JavaVM) -> AttachGuard<'_> {
    let env = jvm.attach_current_thread();
    let thread = call_static_method!(env, {java.lang.Thread.currentThread()});
    let system_class_loader = call_static_method!(env, {java.lang.ClassLoader.getSystemClassLoader()});
    call_method!(env, {void setContextClassLoader(), thread, {system_class_loader}});
    env
}
```

**Comment by chenzl25** [comment]
@wenym1 I finished the refactoring, PTAL again.

File: src/jni_core/src/jvm_runtime.rs [hunk 1]
```diff
@@ -204,11 +204,32 @@ pub fn execute_with_jni_env<T>(
...
+        {ClassLoader},
+        {ClassLoader getSystemClassLoader()}
+    )?;
+
+    crate::call_method!(
+        env,
+        thread,
+        {void setContextClassLoader(ClassLoader)},
+        &system_class_loader
+    )?;
```
Comment (chenzl25) [inline src/jni_core/src/jvm_runtime.rs:232]: `java.lang.Thread.currentThread().setContextClassLoader(java.lang.ClassLoader.getSystemClassLoader());` would be called once after attaching a thread to jvm successfully and any jni related codes should be wrapped in `execute_with_jni_env`.

File: src/connector/src/sink/remote.rs [hunk 2]
```diff
@@ -770,34 +771,31 @@ impl EmbeddedConnectorClient {
...
+                        tracing::error!(error = %e.as_report(), "jni call error");
+                    }
+                };
+
+                Ok(())
+            });
+
+            if let Err(e) = result {
+                let _ = response_tx.blocking_send(Err(e));
+            }
```
Comment (chenzl25) [inline src/connector/src/sink/remote.rs:798]: @wenym1 Do I correctly handle the error in this case?

Comment (wenym1) [reply src/connector/src/sink/remote.rs:798]: LGTM

**Review by wenym1 — APPROVED** [review]
All tests are passed. LGTM

File: src/jni_core/src/jvm_runtime.rs [hunk 3]
```diff
@@ -204,11 +204,32 @@ pub fn execute_with_jni_env<T>(
...
 ) -> anyhow::Result<T> {
-    let _guard = jvm
+    let mut env = jvm
         .attach_current_thread()
         .with_context(|| "Failed to attach current rust thread to jvm")?;
 
-    let mut env = jvm.get_env().with_context(|| "Failed to get jni env")?;
+    // set context class loader for the thread
+    // java.lang.Thread.currentThread()
+    //     .setContextClassLoader(java.lang.ClassLoader.getSystemClassLoader());
```
Comment (StrikeW) [inline src/jni_core/src/jvm_runtime.rs:213]: dead code?

Comment (wenym1) [reply src/jni_core/src/jvm_runtime.rs:213]: This is the equivalent java code, I think it's for reference purpose.

**Review by StrikeW — APPROVED** [review]
LGTM

**Pull request merged**